### PR TITLE
Air bubble

### PR DESCRIPTION
**Pull request by Sindorman**
This PR adds new item - airbubble

The idea is based of [this thread](https://forums.aurorastation.org/viewtopic.php?f=18&t=5739).

- Adds airbubble - item which when unfolded can provide user with environmental protection, the pressure inside won't be idea, but the oxygen will be enough to last for 20 minutes.

Airbubble properties:

1. When factory folded it is of normal size, so that it can fit in your internals box or backpack. But when folded manually after usage it is large size, and can only fit in storage like backpack minimal.

2. Has internal tank attached to it. By default each air bubble comes with double emergency oxygen tank(which can be found in engineering), it has enough oxygen for 20 minutes of usage. Tank can be detaches to be refilled and attach back. You can also attach different kind of tanks and they would work. Example is to attach phoron tank to kill someone, or sleep toxin to make them from breaking out.

3. Airbubble zipper can be forcefully blocked by wire handcuffs, to kidnap people. To take it out you need to use wirecutters.

4. Provides with breathable atmosphere, but pressure is just above minimal - meaning that you won't feel that comfortable inside of it

5. Breaking out when forcefully blocked is possible using resist, it would though make a hole in the bubble making it leak air of tank valve is on and not possible to use to protect yourself from environment.

TODO list:

- [x] Clean unused code

- [x] Add sprites for air bubble

- [x] Add airbubble to every emergency internals locker

### CODE REVIEW

File: code/game/objects/items/airbubble.dm [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+	if(use_internal_tank)
+		. = inside_air.temperature
+	else
+		var/datum/gas_mixture/t_air = get_turf_air()
+		if(t_air)
+			. = t_air.temperature
+	return
+
+/obj/structure/closet/air_bubble/process()
+	var/static/max_ticks = 16
```
Comment (Lohikar) [inline code/game/objects/items/airbubble.dm]: Why isn't this a constant or define?

File: code/game/objects/items/airbubble.dm [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+	close_sound = 'sound/items/zip.ogg'
+	var/item_path = /obj/item/air_bubble
+	var/zipped = 0
+	density = 0
+	storage_capacity = 30
+	var/contains_body = 0
+	var/used = 1
+
+	var/use_internal_tank = 1
+	var/current_processes = BUBBLE_PROC_INT_TEMP
```
Comment (Lohikar) [inline code/game/objects/items/airbubble.dm]: Why isn't this a boolean? It only has two states.

File: code/game/objects/items/airbubble.dm [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+		inside_air.temperature -= max(-10, min(10, round(delta/4,0.1)))
+
+/////////////////////////////////////////
+//////// airbubble process() helpers ////
+/////////////////////////////////////////
+
+/obj/structure/closet/air_bubble/proc/stop_process(process)
+	current_processes &= ~process
+
+/obj/structure/closet/air_bubble/proc/start_process(process)
```
Comment (Lohikar) [inline code/game/objects/items/airbubble.dm]: This doesn't seem to be used anywhere.

File: code/game/objects/items/airbubble.dm [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+/obj/structure/closet/air_bubble/proc/process_preserve_temp()
+	if (inside_air && inside_air.volume > 0)
+		var/delta = inside_air.temperature - T20C
+		inside_air.temperature -= max(-10, min(10, round(delta/4,0.1)))
+
+/////////////////////////////////////////
+//////// airbubble process() helpers ////
+/////////////////////////////////////////
+
+/obj/structure/closet/air_bubble/proc/stop_process(process)
```
Comment (Lohikar) [inline code/game/objects/items/airbubble.dm]: This doesn't seem to be used anywhere.

File: code/game/objects/items/airbubble.dm [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+	else
+		var/turf/T = get_turf(src)
+		if(T)
+			return T.remove_air(amount)
+	return
+
+/obj/structure/closet/air_bubble/return_air()
+	if(use_internal_tank)
+		return inside_air
+	return get_turf_air()
```
Comment (Lohikar) [inline code/game/objects/items/airbubble.dm]: Why not just call parent? Parent returns turf's air pretty sure.

File: code/game/objects/items/airbubble.dm [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+/obj/item/air_bubble
+	name = "air bubble"
+	desc = "Special air bubble designed to protect people inside of it from decompressed enviroments."
+	icon = 'icons/obj/bodybag.dmi'
+	icon_state = "bodybag_folded"
+	w_class = ITEMSIZE_SMALL
+	var/used = 0
+	var/obj/item/weapon/tank/emergency_oxygen/double/internal_tank
+
+	attack_self(mob/user)
```
Comment (Lohikar) [inline code/game/objects/items/airbubble.dm]: Absolute pathing please.

File: code/game/objects/items/oxycandle.dm [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	desc = "Oxygen candles are flare-like steel tubes of sodium chlorate that produces oxygen through chemical process."
+	icon = 'icons/obj/candle.dmi'
+	icon_state = "candle1"
+	w_class = ITEMSIZE_SMALL
+	var/target_pressure = ONE_ATMOSPHERE
+	var/datum/gas_mixture/air_contents = null
+	var/volume = 600
+	var/on = 0
+
+	attack_self(mob/user)
```
Comment (Lohikar) [inline code/game/objects/items/oxycandle.dm]: Absolute pathing please.

File: code/game/objects/items/oxycandle.dm [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		icon_state = "candle1_lit"
+		update_icon()
+		air_contents = new /datum/gas_mixture()
+		air_contents.volume = volume //liters
+		air_contents.temperature = T20C
+		var/list/air_mix = StandardAirMix()
+		src.air_contents.adjust_multi("oxygen", air_mix["oxygen"], "nitrogen", air_mix["nitrogen"])
+		START_PROCESSING(SSprocessing, src)
+
+/obj/item/oxycandle/proc/StandardAirMix()
```
Comment (Lohikar) [inline code/game/objects/items/oxycandle.dm]: Why is this needed?

File: code/game/objects/items/oxycandle.dm [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+/obj/item/oxycandle/proc/MolesForPressure()
+	return (target_pressure * air_contents.volume) / (R_IDEAL_GAS_EQUATION * air_contents.temperature)
+
+/obj/item/oxycandle/process()
+	if(!on)
+		return
+	else
+		if(!volume)
+			new/obj/item/trash/candle(src.loc)
+			if(istype(src.loc, /mob))
```
Comment (Lohikar) [inline code/game/objects/items/oxycandle.dm]: `ismob(loc)`

File: code/game/objects/items/airbubble.dm (hunk 1 above)
Comment (Sindorman) [reply code/game/objects/items/airbubble.dm]: Idk, It was borrowed from Mech's code, which doesn't have it.

File: code/game/objects/items/airbubble.dm (hunk 4 above)
Comment (Sindorman) [reply code/game/objects/items/airbubble.dm]: True, can be deleted.

File: code/game/objects/items/airbubble.dm (hunk 6 above)
Comment (Sindorman) [reply code/game/objects/items/airbubble.dm]: What do you mean?

File: code/game/objects/items/oxycandle.dm (hunk 9 above)
Comment (Sindorman) [reply code/game/objects/items/oxycandle.dm]: Ok

**Comment by Sindorman** [comment]
I did not base it off Bugfix PR, I accidentally rebased it on top of master.

**Comment by Sindorman** [comment]
I have cleaned PR from dirt

**Comment by Sindorman** [comment]
This is no longer WIP, finished. Review required.

**Comment by BotBOREALIS** [comment]
Icons with diff:
<details><summary>icons/obj/airbubble.dmi</summary>

Key | Old | New | Status
--- | --- | --- | ---
airbubble||airbubble|Created
airbubble_open||airbubble_open|Created
airbubble_fact_folded||airbubble_fact_folded|Created
airbubble_man_folded||airbubble_man_folded|Created
airbubble_syndie||airbubble_syndie|Created
airbubble_syndie_open||airbubble_syndie_open|Created
airbubble_syndie_fact_folded||airbubble_syndie_fact_folded|Created
airbubble_syndie_man_folded||airbubble_syndie_man_folded|Created
</details>

**Comment by BotBOREALIS** [comment]
Icons with diff:
<details><summary>icons/obj/airbubble.dmi</summary>

Key | Old | New | Status
--- | --- | --- | ---
airbubble||airbubble|Created
airbubble_open||airbubble_open|Created
airbubble_fact_folded||airbubble_fact_folded|Created
airbubble_syndie||airbubble_syndie|Created
airbubble_syndie_open||airbubble_syndie_open|Created
airbubble_syndie_fact_folded||airbubble_syndie_fact_folded|Created
airbubble_man_folded||airbubble_man_folded|Created
airbubble_syndie_man_folded||airbubble_syndie_man_folded|Created
</details>

File: code/game/objects/items/airbubble.dm [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+			qdel(src)
+			return
+
+/obj/structure/closet/airbubble/verb/set_internals(mob/user as mob)
+	set src in oview(1)
+	set category = "Object"
+	set name = "Set Internals"
+	if(use_internal_tank)
+		visible_message(
+			"<span class='warning'>[user] set [src] internals.</span>",
```
Comment (Arrow768) [inline code/game/objects/items/airbubble.dm]: Shouldnt it be is setting / you are trying to set ? -> Since you are doing it right now (before the do_after)

File: code/game/objects/items/airbubble.dm [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+
+	playsound(loc, close_sound, 25, 0, -3)
+	density = 1
+	add_inside()
+	return 1
+
+/obj/structure/closet/airbubble/MouseDrop(over_object, src_location, over_location)
+	..()
+	if(!zipped)
+		if((over_object == usr && (in_range(src, usr) || usr.contents.Find(src))))
```
Comment (Arrow768) [inline code/game/objects/items/airbubble.dm]: 116 and 117 could be combined into one line

File: code/game/objects/items/airbubble.dm [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+	else
+		visible_message("<span class='notice'>[src] has no internal tank.</span>")
+
+/obj/structure/closet/airbubble/verb/take_tank(mob/user as mob)
+	if(use_internal_tank)
+		set src in oview(1)
+		set category = "Object"
+		set name = "Remove Air Tank"
+		visible_message(
+		"<span class='warning'>[user] removed [internal_tank] from [src].</span>",
```
Comment (Arrow768) [inline code/game/objects/items/airbubble.dm]: Similar here with is removing / you are trying to remove

File: code/game/objects/items/airbubble.dm [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+		STOP_PROCESSING(SSfast_process, src)
+	else
+		visible_message(
+		"<span class='warning'>[src] already has no tank.</span>")
+
+/obj/structure/closet/airbubble/attackby(W as obj, mob/user as mob)
+	if(istype(W, /obj/item/weapon/tank))
+		if(!use_internal_tank)
+			user.visible_message(
+				"<span class='warning'>[user] attached [W] to [src].</span>",
```
Comment (Arrow768) [inline code/game/objects/items/airbubble.dm]: again, is attaching / you are trying to attach

File: code/game/objects/items/airbubble.dm [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+			MouseDrop_T(G.affecting, user)
+			return 0
+		if(istype(W,/obj/item/tk_grab))
+			return 0
+		if(!dropsafety(W))
+			return
+		usr.drop_item()
+	else if(istype(W, /obj/item/weapon/handcuffs/cable))
+		user.visible_message(
+			"<span class='warning'>[user] begins putting cable restrains on zipper of [src].</span>",
```
Comment (Arrow768) [inline code/game/objects/items/airbubble.dm]: Shouldnt it be "on the zipper of (the)"

File: code/game/objects/items/airbubble.dm [hunk 15]
```diff
@@ -0,0 +1,462 @@
...
+	w_class = ITEMSIZE_NORMAL
+	var/used = FALSE
+	var/ripped = FALSE
+	var/obj/item/weapon/tank/internal_tank
+
+/obj/item/airbubble/Destroy()
+	qdel(internal_tank)
+	return ..()
+
+/obj/item/airbubble/attack_self(mob/user)
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: Put a `if (!isturf(user.loc)) to_chat(user, "You're fucking stupid, the air bubble can't deploy in an enclosed space.")` here.

Otherwise you open it in a locker and then you have a locker in a locker and aaaaaaaaaaaaa.

File: code/game/objects/items/airbubble.dm [hunk 16]
```diff
@@ -0,0 +1,462 @@
...
+	var/obj/structure/closet/airbubble/R = new /obj/structure/closet/airbubble(user.loc)
+	if(!used && !ripped)
+		internal_tank = new /obj/item/weapon/tank/emergency_oxygen/double(src)
+	R.internal_tank = internal_tank
+	internal_tank.forceMove(R)
+	internal_tank = null
+	R.add_fingerprint(user)
+	R.ripped = ripped
+	R.update_icon()
+	qdel(src)
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: Shouldn't you also do a `user.drop_from_inventory(src)`? Otherwise a ghost object may persist.

File: code/game/objects/items/airbubble.dm [hunk 17]
```diff
@@ -0,0 +1,462 @@
...
+/obj/structure/closet/airbubble/can_open()
+	if(zipped)
+		return 0
+	return 1
+
+/obj/structure/closet/airbubble/can_close()
+	if(zipped)
+		return 0
+	var/turf/T = get_turf(src)
+	for(var/obj/structure/closet/closet in T)
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: This can be generalized.

```dm
for (var/obj/O in T)
    if ((O.density || O.anchored) && O != src)
        return 0
```

File: code/game/objects/items/airbubble.dm [hunk 18]
```diff
@@ -0,0 +1,462 @@
...
+		mob_num += 1
+		if(mob_num > 1)
+			visible_message("<span class='warning'>[src] can only fit one person.</span>")
+			return 0
+	return 1
+
+/obj/structure/closet/airbubble/dump_contents()
+
+	for(var/obj/I in src)
+		if(!istype(I, /obj/item/weapon/tank))
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: Bad idea, you might legitimately have multiple air tanks in the `src`. Do this instead:

```
if (I != internal_tank)
    I.forceMove(loc)
```

File: code/game/objects/items/airbubble.dm [hunk 19]
```diff
@@ -0,0 +1,462 @@
...
+			STOP_PROCESSING(SSfast_process, src)
+		else
+			START_PROCESSING(SSfast_process, src)
+		use_internal_tank = !use_internal_tank
+	else
+		visible_message("<span class='notice'>[src] has no internal tank.</span>")
+
+/obj/structure/closet/airbubble/verb/take_tank(mob/user as mob)
+	if(use_internal_tank)
+		set src in oview(1)
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: This form of syntax is generally not used.  The `set` directives are almost always int he header.

File: code/game/objects/items/airbubble.dm [hunk 20]
```diff
@@ -0,0 +1,462 @@
...
+	if(opened)
+		if(istype(W, /obj/item/weapon/grab))
+			var/obj/item/weapon/grab/G = W
+			MouseDrop_T(G.affecting, user)
+			return 0
+		if(istype(W,/obj/item/tk_grab))
+			return 0
+		if(!dropsafety(W))
+			return
+		usr.drop_item()
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: `usr` => `user`

File: code/game/objects/items/airbubble.dm (hunk 15 above)
Comment (Sindorman) [reply code/game/objects/items/airbubble.dm]: That one is solved

File: code/game/objects/items/airbubble.dm [hunk 21]
```diff
@@ -0,0 +1,492 @@
...
+		STOP_PROCESSING(SSfast_process, src)
+	else
+		visible_message("<span class='warning'>[src] already has no tank.</span>")
+
+// Handle most of things: restraining, cutting restrains, attaching tank.
+/obj/structure/closet/airbubble/attackby(W as obj, mob/user as mob)
+	if(istype(W, /obj/item/weapon/tank))
+		if(!use_internal_tank)
+			visible_message("<span class='warning'>[user] is attaching [W] to [src].</span>")
+			user << "<span class='notice'>You are attaching [W] to [src].</span>"
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: Do this instead:
```dm
user.visible_message("<span class='warning'>[user] is attaching [W] to [src].</span>", "<span class='notice'>You are attaching [W] to [src].</span>")
```

The second argument is passed to the caller.

File: code/game/objects/items/airbubble.dm [hunk 22]
```diff
@@ -0,0 +1,492 @@
...
+			if (!do_after(user, 2 SECONDS, act_target = src))
+				return
+			visible_message("<span class='warning'>[user] have attached [W] to [src].</span>")
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: Do as described above.

File: code/game/objects/items/airbubble.dm [hunk 23]
```diff
@@ -0,0 +1,492 @@
...
+	else
+		visible_message("<span class='notice'>[src] has no internal tank.</span>")
+
+// Remove tank from bubble
+/obj/structure/closet/airbubble/verb/take_tank(mob/user as mob)
+	set src in oview(1)
+	set category = "Object"
+	set name = "Remove Air Tank"
+	if(!isnull(internal_tank))
+		visible_message("<span class='warning'>[user] is removing [internal_tank] from [src].</span>")
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: See note about visible_message followed by user.

File: code/game/objects/items/airbubble.dm [hunk 24]
```diff
@@ -0,0 +1,492 @@
...
+		user << "<span class='notice'>You are removing [internal_tank] from [src].</span>"
+		if (!do_after(user, 2 SECONDS, act_target = src))
+			return
+		visible_message("<span class='warning'>[user] have removed [internal_tank] from [src].</span>")
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: See note about visible_message followed by user.

File: code/game/objects/items/airbubble.dm [hunk 25]
```diff
@@ -0,0 +1,492 @@
...
+		return 0
+	var/turf/T = get_turf(src)
+	for (var/obj/O in T)
+		if (O.density && O != src)
+			return 0
+	var/mob_num = 0
+	for(var/mob/living/M in T)
+		mob_num += 1
+		if(mob_num > 1)
+			visible_message("<span class='warning'>[src] can only fit one person.</span>")
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: This seems really weird. It'll print a visible message? Use another way of phrasing it. `[src] is overly full and won't close!`

File: code/game/objects/items/airbubble.dm [hunk 26]
```diff
@@ -0,0 +1,492 @@
...
+	escapee.next_move = world.time + 100
+	escapee.last_special = world.time + 100
+	escapee << "<span class='warning'>You lean on the back of \the [src] and start punching internal wall with your legs. (this will take about [breakout_time] minutes)</span>"
+	visible_message("<span class='danger'>\The [src] begins to shake violently! Something is terring it from the inside!</span>")
+
+	var/time = 6 * breakout_time * 2
+
+	var/datum/progressbar/bar
+	if (escapee.client && escapee.client.prefs.toggles_secondary & PROGRESS_BARS)
+		bar = new(escapee, time, src)
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: Is this level of reinvention necessary? `do_after` has an argument `extra_checks`. It can be used to do what's required. It eager-returns if they fail. So basically:

```dm
/obj/structure/closet/airbubble/proc/breakout_callback(/mob/living/escapee)
	if (QDELETED(escapee))
		return FALSE

	if ((world.time - last_shake) > 5 SECONDS)
		playsound(loc, "sound/items/[pick("rip1","rip2")].ogg", 100, 1)
		animate_shake()
		last_shake = world.time

	if (!req_breakout())
		return FALSE

	return TRUE

/obj/structure/closet/airbubble/mob_breakout(/mob/living/escapee)
	if (req_breakout() < 1)
		return

	escapee.next_move = world.time + 100
	escapee.last_special = world.time + 100
	escapee << "<span class='warning'>You lean on the back of \the [src] and start punching internal wall with your legs. (this will take about [breakout_time] minutes)</span>"
	visible_message("<span class='danger'>\The [src] begins to shake violently! Something is terring it from the inside!</span>")

	var/time = 6 * breakout_time * 2
	breakout = TRUE

	if (!do_after(escapee, time, extra_checks = CALLBACK(src, .proc/breakout_callback, escapee)))
		breakout = FALSE
		return

	breakout = FALSE
	escapee << "<span class='warning'>You successfully break out! Tearing the bubble's walls!</span>"
	visible_message("<span class='danger'>\the [escapee] successfully broke out of \the [src]! Tearing the bubble's walls!</span>")
	playsound(loc, "sound/items/[pick("rip1","rip2")].ogg", 100, 1)
	break_open()
	animate_shake()
	qdel(bar)
```

File: code/game/objects/items/airbubble.dm [hunk 27]
```diff
@@ -0,0 +1,492 @@
...
+			return
+		visible_message("<span class='warning'>[user] have set [src] internals.</span>")
+		user << "<span class='notice'>You set [src] internals.</span>"
+		if(use_internal_tank)
+			STOP_PROCESSING(SSfast_process, src)
+		else
+			START_PROCESSING(SSfast_process, src)
+		use_internal_tank = !use_internal_tank
+	else
+		visible_message("<span class='notice'>[src] has no internal tank.</span>")
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: Why is this a visible message? Display it to the user instead.

File: code/game/objects/items/airbubble.dm [hunk 28]
```diff
@@ -0,0 +1,492 @@
...
+			var/obj/item/weapon/grab/G = W
+			MouseDrop_T(G.affecting, user)
+			return 0
+		if(istype(W,/obj/item/tk_grab))
+			return 0
+		if(!dropsafety(W))
+			return
+		user.drop_item()
+	else if(istype(W, /obj/item/weapon/handcuffs/cable))
+		visible_message("<span class='warning'>[user] begins putting cable restrains on zipper of [src].</span>")
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: visible_message -> `user.visible_message("first", "second")`

File: code/game/objects/items/airbubble.dm [hunk 29]
```diff
@@ -0,0 +1,492 @@
...
+		user << "<span class='notice'>You begin putting cable restrains on zipper of [src].</span>"
+		playsound(loc, 'sound/weapons/cablecuff.ogg', 50, 1)
+		if (!do_after(user, 3 SECONDS, act_target = src, extra_checks = CALLBACK(src, .proc/is_closed)))
+			return
+		zipped = !zipped
+		update_icon()
+		user.visible_message("<span class='warning'>[src]'s zipper has been zipped by [user].</span>")
+		user << "<span class='notice'>You put restrains on [src]'s zipper.</span>"
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: Put this text as the second argumetn to the line above it.

File: code/game/objects/items/airbubble.dm [hunk 30]
```diff
@@ -0,0 +1,492 @@
...
+
+		qdel(W)
+		update_icon()
+	else if(istype(W, /obj/item/weapon/wirecutters))
+		visible_message("<span class='warning'>[user] begins cutting cable restrains on zipper of [src].</span>")
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: `visible_message("first") + user << "second"` -> `user.visible_message("first", "second")`

File: code/game/objects/items/airbubble.dm [hunk 31]
```diff
@@ -0,0 +1,492 @@
...
+		user << "<span class='notice'>You begin cutting cable restrains on zipper of [src].</span>"
+		playsound(loc, 'sound/items/Wirecutter.ogg', 50, 1)
+		if (!do_after(user, 3 SECONDS, act_target = src, extra_checks = CALLBACK(src, .proc/is_closed)))
+			return
+		zipped = !zipped
+		update_icon()
+		visible_message("<span class='warning'>[src] zipper's cable restrains have been cut by [user].</span>")
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: `visible_message("first") + user << "second"` -> `user.visible_message("first", "second")`

File: code/game/objects/items/airbubble.dm [hunk 32]
```diff
@@ -208,72 +222,88 @@
...
+		user.visible_message(
+		"<span class='warning'>[user] is setting [src] internals.</span>",
+		"<span class='notice'>You are settting [src] internals.</span>"
+		)
 		if (!do_after(user, 2 SECONDS, act_target = src))
 			return
-		visible_message("<span class='warning'>[user] have set [src] internals.</span>")
-		user << "<span class='notice'>You set [src] internals.</span>"
+		user.visible_message(
+		"<span class='warning'>[user] have set [src] internals.</span>" ,
```
Comment (Arrow768) [inline code/game/objects/items/airbubble.dm]: shouldnt it be has set ?

File: code/game/objects/items/airbubble.dm [hunk 33]
```diff
@@ -208,72 +222,88 @@
...
+		user.visible_message(
+		"<span class='warning'>[user] is removing [internal_tank] from [src].</span>",
+		"<span class='notice'>You are removing [internal_tank] from [src].</span>"
+		)
 		if (!do_after(user, 2 SECONDS, act_target = src))
 			return
-		visible_message("<span class='warning'>[user] have removed [internal_tank] from [src].</span>")
-		user << "<span class='notice'>You removed [internal_tank] from [src].</span>"
+		user.visible_message(
+		"<span class='warning'>[user] have removed [internal_tank] from [src].</span>",
```
Comment (Arrow768) [inline code/game/objects/items/airbubble.dm]: "has removed" / "removed" ?

File: code/game/objects/items/airbubble.dm (hunk 32 above)
Comment (Sindorman) [reply code/game/objects/items/airbubble.dm]: Well, I was assuming we use pronoun `they`?

File: code/game/objects/items/airbubble.dm [hunk 34]
```diff
@@ -208,72 +222,88 @@
...
+			user.visible_message(
+			"<span class='warning'>[user] is attaching [W] to [src].</span>",
+			"<span class='notice'>You are attaching [W] to [src].</span>"
+			)
 			if (!do_after(user, 2 SECONDS, act_target = src))
 				return
-			visible_message("<span class='warning'>[user] have attached [W] to [src].</span>")
-			user << "<span class='notice'>You attached [W] to [src].</span>"
+			user.visible_message(
+			"<span class='warning'>[user] have attached [W] to [src].</span>",
```
Comment (Arrow768) [inline code/game/objects/items/airbubble.dm]: "has attached" / "attached" ?

**Comment by Sindorman** [comment]
@skull132 Finished!

File: code/game/objects/items/airbubble.dm [hunk 35]
```diff
@@ -0,0 +1,479 @@
...
+// We are out finally, the bubble is ripped. So dump everything out from it. Especially air and user.
+/obj/structure/closet/airbubble/break_open()
+	ripped = TRUE
+	update_icon()
+	dump_contents()
+	var/datum/gas_mixture/t_air = get_turf_air()
+	t_air.merge(inside_air)
+
+// Change valve on internal tank
+/obj/structure/closet/airbubble/verb/set_internals(mob/user as mob)
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: I should have probably noticed this before. This verb should not take arguments. The verb activator (usr) is not implicitly the first argument. If used, what happens is a menu is given to the usr to select who the user is, _if_ there are multiple candidates. eg. multiple mobs around.

SO: remove the argument for the verb. And replace internal references to `user` with `usr`. This is the one set of cases where `usr` is valid: in verbs.

File: code/game/objects/items/airbubble.dm [hunk 36]
```diff
@@ -0,0 +1,479 @@
...
+		if(use_internal_tank)
+			STOP_PROCESSING(SSfast_process, src)
+		else
+			START_PROCESSING(SSfast_process, src)
+		use_internal_tank = !use_internal_tank
+	else
+		user << "<span class='notice'>[src] has no internal tank.</span>"
+
+// Remove tank from bubble
+/obj/structure/closet/airbubble/verb/take_tank(mob/user as mob)
```
Comment (skull132) [inline code/game/objects/items/airbubble.dm]: Same here. Remove argument, replace references to `user` with `usr`.
